core/runtime/common/runtime_upgrade_tracker_impl.cpp

```diff
@@ -9,27 +9,43 @@
 #include "blockchain/block_tree.hpp"
 #include "log/profiling_logger.hpp"
 #include "runtime/common/storage_code_provider.hpp"
+#include "storage/predefined_keys.hpp"
 
 namespace kagome::runtime {
 
   RuntimeUpgradeTrackerImpl::RuntimeUpgradeTrackerImpl(
-      std::shared_ptr<const blockchain::BlockHeaderRepository> header_repo)
+      std::shared_ptr<const blockchain::BlockHeaderRepository> header_repo,
+      std::shared_ptr<storage::BufferStorage> storage,
+      const application::CodeSubstitutes &code_substitutes)
       : header_repo_{std::move(header_repo)},
+        storage_{std::move(storage)},
+        code_substitutes_{code_substitutes},
         logger_{log::createLogger("StorageCodeProvider", "runtime")} {
     BOOST_ASSERT(header_repo_);
+    BOOST_ASSERT(storage_);
+    auto encoded_res = storage_->get(storage::kRuntimeHashesLookupKey);
+    if (encoded_res.has_value()) {
+      auto decoded_res = scale::decode<decltype(runtime_upgrade_parents_)>(
+          encoded_res.value());
+      if (not decoded_res.has_value()) {
+        SL_ERROR(logger_, "Saved runtime hashes data structure is incorrect!");
+      } else {
+        runtime_upgrade_parents_ = decoded_res.value();
+      }
+    }
   }
 
-  outcome::result<storage::trie::RootHash>
-  RuntimeUpgradeTrackerImpl::getLastCodeUpdateState(
-      const primitives::BlockInfo &block) const {
+  outcome::result<primitives::BlockId>
```

Comment:
RootHash still has to be returned to avoid additional expensive calls to header repo.